完善骨骼动画相关 (#1553)
cocos-creator/creator-docs: en/editor/components/spine.md

Select the node and choose **Add Component -> Spine -> Skeleton** on the **Inspector** panel to add the Skeleton component to the node.

- For the use of the Spine Skeleton component, please refer to the [Spine Skeleton example case](https://github.com/cocos-creator/test-cases-3d/tree/v3.0/assets/cases/spine) for details.
- For Spine Skeleton component related scripting interfaces, please refer to the [Spine Skeleton API](__APIDOC__/en/classes/spine.skeleton-1.html) for details.

## Spine Properties

| Property | Description
| :-------------------- | :----------------- |
| CustomMaterial        | Custom materials that can be used to achieve rendering effects such as dissolve, external glow, etc. Please refer to the [Custom Materials](../../ui-system/components/engine/ui-material.md) for details.
| Color                 | Set the skeleton animation color.
| SkeletonData          | Skeleton information data, drag and drop the skeleton assets exported from Spine. into this property
| Default Skin          | Select the default skin.
| Animation             | The name of the currently playing animation.

> | Animation Cache Mode  | Rendering mode, the default is **REALTIME** mode. <br>1. **REALTIME** mode, real-time computing, supports all Spine features. <br>2. **SHARED_CACHE** mode, caches and shares skeletal animations and texture data, equivalent to pre-baked skeletal animations. Has higher performance, but does not support motion fusion, motion overlay, and only supports motion start and end events. As for memory, when creating N (N>=3) animations with the same skeleton and the same action, the memory advantage is obvious. In summary, **SHARED_CACHE** mode is suitable for scene animations, effects, replica monsters, NPCs, etc., and can greatly improve frame rates and reduce memory consumption. <br>3. **PRIVATE_CACHE** mode, similar to **SHARED_CACHE** but does not share animation and texture data, so there is no memory advantage, only performance advantage. If you want to take advantage of the high performance of the cache mode, but also want to realize the dressing function (the texture data cannot be shared), then **PRIVATE_CACHE** is for you.

Reviewer: what is "dressing function" as used in the original text?

> | Animation Cache Mode  | Rendering mode, the default is **REALTIME** mode. <br>1. **REALTIME** mode, real-time computing, supports all Spine features. <br>2. **SHARED_CACHE** mode, caches and shares skeletal animations and texture data, equivalent to pre-baked skeletal animations. Has higher performance, but does not support motion fusion, motion overlay, and only supports motion start and end events. As for memory, when creating N (N>=3) animations with the same skeleton and the same action, the memory advantage is obvious. In summary, **SHARED_CACHE** mode is suitable for scene animations, effects, replica monsters, NPCs, etc., and can greatly improve frame rates and reduce memory consumption. <br>3. **PRIVATE_CACHE** mode, similar to **SHARED_CACHE** but does not share animation and texture data, so there is no memory advantage, only performance advantage. If you want to take advantage of the high performance of the cache mode, but also want to realize the dressing function (the texture data cannot be shared), then **PRIVATE_CACHE** is for you.

Author: change the texture of Spine

        start () {
            this._jitterEffect = new sp.VertexEffectDelegate();
            // Set the jitter parameters.
            this._jitterEffect.initJitter(20, 20);
            // Call the 'setVertexEffectDelegate' method of the Skeleton component to set the effect.
            this.skeleton!.setVertexEffectDelegate(this._jitterEffect!);
        }

    };
    ```

3. Next, attach the `SpineExample` script to the `Canvas` node, i.e. drag and drop the script into the node's **Inspector** panel. Then drag the `Spine` node with the **Skeleton** component attached in the **Hierarchy** panel to the corresponding `Skeleton` property box of the script component, and save the scene.

4. Click the **Preview** button at the top of the editor to see the effect of vertex jitter of the Spine animation. For the code, please refer to the [SpineMesh](https://github.com/cocos-creator/test-cases-3d/tree/v3.0/assets/cases/spine) example.

## Spine Attachment

When using skeleton animation, nodes are often attached on a certain part of the skeleton animation to achieve the effect of linkage between the nodes and the skeleton animation.

> We can implement the Spine Attachments by using both editor and script. Here is an example of how Spine uses attachment to attach a star to the dragon's tail, and shake it with the dragon's tail.

Reviewer: ```suggestion
Spine Attachments can be implemented by using both editor and script. Here is an example of how Spine uses an attachment to attach a star to the dragon's tail, and shake it with the dragon's tail.
```

![attach0](./spine/attach0.png)

### Implemented via the editor

> 1. First create a new `Canvas` node in the **Hierarchy** panel, and then create a new empty node and name it to `Spine` under the `Canvas` node. Select `Spine` and add the **Skeleton** component to the **Inspector** panel. Drag and drop the asset into the `SkeletonData` property box of the **Skeleton** component and set the Skeleton component properties.

Reviewer: ```suggestion
1. First, create a new `Canvas` node in the **Hierarchy** panel, and then create a new empty node and name it to `Spine` under the `Canvas` node. Select `Spine` and add the **Skeleton** component to the **Inspector** panel. Drag and drop the asset into the `SkeletonData` property box of the **Skeleton** component and set the Skeleton component properties.
```

> 2. Then right-click on the Spine node in the **Hierarchy** panel and select **Create -> 2D Objects -> Sprite** to add a child node for it and name it `star`. Then drag the star asset to the `SpriteFrame` property of the Sprite component in the **Inspector** panel.

Reviewer: ```suggestion
2. Second, right-click on the Spine node in the **Hierarchy** panel and select **Create -> 2D Objects -> Sprite** to add a child node for it and name it `star`. Drag the star asset to the `SpriteFrame` property of the Sprite component in the **Inspector** panel.
```

    ![attach1](./spine/attach1.png)

3. Select the Spine node in the **Hierarchy** panel, and set the Skeleton component's `Sockets` property to **1** in the **Inspector** panel (the value of the `Sockets` property represents the number of attachments).

    ![attach2](./spine/attach2.png)

> 4. Next, set the `Path` and `Target` properties of the `Sockets`. The `Path` drop-down box will list all the skeletons, select the target bone you want to attach, here take the dragon's tail as an example, then drag the `star` node to the `Target` property box. You can see the star attached on the dragon's tail in the **Scene** panel.

Reviewer: ```suggestion
4. Next, set the `Path` and `Target` properties of the `Sockets`. The `Path` drop-down box will list all the skeletons, select the target bone you want to attach, here take the dragon's tail as an example, then drag the `star` node to the `Target` property box. Notice the star attached on the dragon's tail in the **Scene** panel.
```

    ![attach3](./spine/attach3.png)

> 5. Finally, Save the scene and click the **Preview** button on top of the editor to see the star hanging on the dragon's tail and swaying along with it. Please refer to the [SpineAttach](https://github.com/cocos-creator/test-cases-3d/tree/v3.0/assets/cases/spine) example for details.

Reviewer: ```suggestion
5. Finally, save the scene and click the **Preview** button on top of the editor to see the star hanging on the dragon's tail and swaying along with it. Please refer to the [SpineAttach](https://github.com/cocos-creator/test-cases-3d/tree/v3.0/assets/cases/spine) example for details.
```

> ### Implemented via code

Reviewer: ```suggestion
### Implementing via code
```

1. The first two steps are the same as those implemented via the editor.

2. Create a new TypeScript script in the Explorer panel and name it `SpineAttach`. Write the component script. The script code is as follows:

    ```ts
    import { _decorator, Component, sp, Label, Node } from 'cc';
    const { ccclass, property } = _decorator;

    @ccclass('SpineAttach')
    export class SpineAttach extends Component {

        @property({ type: sp.Skeleton })
        skeleton: sp.Skeleton = null!;

        @property({ type: Node })
        attachNode: Node = null!;

        start () {
            var socket = new sp.SpineSocket("root/hip/tail1/tail2/tail3/tail4/tail5/tail6/tail7/tail8/tail9/tail10", this.attachNode); // The first incoming parameter is the target bone of the attachment, and the second incoming parameter is the node of the attachment.
            this.skeleton!.sockets.push(socket);
            this.skeleton!.sockets = this.skeleton!.sockets;
        }

    }
    ```

>     If you don't know the name of the target bone, set the `Sockets` property in the Skeleton component to **1** and then look for the name of the desired target bone in the `Path` drop-down box. When the search is complete, restore the `Sockets` property to 0.

Reviewer: ```suggestion
    If the name of the target bone is unknown, set the `Sockets` property in the Skeleton component to **1** and then look for the name of the desired target bone in the `Path` drop-down box. When the search is complete, restore the `Sockets` property to 0.
```

> 3. Next, attach the `SpineAttach` script to the `Canvas` node, i.e. drag and drop the script into the node's **Inspector** panel. Then drag and drop the `Spine` node with the Skeleton component attached and the `star` node in the **Hierarchy** panel to the `Skeleton` property box and the `AttachNode` property box of the script component, respectively, and save the scene.

Reviewer: ```suggestion
3. Next, attach the `SpineAttach` script to the `Canvas` node, i.e. drag and drop the script into the node's **Inspector** panel. Drag and drop the `Spine` node with the Skeleton component attached and the `star` node in the **Hierarchy** panel to the `Skeleton` property box and the `AttachNode` property box of the script component, respectively, and save the scene.
```
